tools/testbench/common_test_ipc4.c

```diff
@@ -0,0 +1,593 @@
+			ret = tb_free_widgets(tp, starting_comp_info, route_info->sink);
+			if (ret < 0)
+				return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int tb_free_widgets_capture(struct testbench_prm *tp,
+				   struct tplg_comp_info *starting_comp_info,
+				   struct tplg_comp_info *current_comp_info)
+{
+	struct tplg_route_info *route_info;
+	struct list_item *item;
+	int ret;
+
+	/* for playback */
+	list_for_item(item, &tp->route_list) {
+		route_info = container_of(item, struct tplg_route_info, item);
+		if (route_info->sink != current_comp_info)
+			continue;
+
+		/* Widgets will be freed when the pipeline is deleted, so just unbind modules */
+		ret = tb_free_route(tp, route_info);
+		if (ret < 0)
+			return ret;
+
+		/* and then continue down the path */
+		if (route_info->sink->type != SND_SOC_TPLG_DAPM_DAI_IN &&
+		    route_info->sink->type != SND_SOC_TPLG_DAPM_DAI_OUT) {
+			ret = tb_free_widgets_capture(tp, starting_comp_info, route_info->source);
+			if (ret < 0)
+				return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int tb_free_pipelines(struct testbench_prm *tp, int dir)
+{
+	struct tplg_pipeline_list *pipeline_list;
+	struct tplg_pcm_info *pcm_info;
+	struct list_item *item;
+	struct tplg_comp_info *host = NULL;
+	int ret, i;
+
+	list_for_item(item, &tp->pcm_list) {
+		pcm_info = container_of(item, struct tplg_pcm_info, item);
+		if (dir)
+			host = pcm_info->capture_host;
+		else
+			host = pcm_info->playback_host;
+
+		if (!host || !tb_is_pipeline_enabled(tp, host->pipeline_id))
+			continue;
+
+		if (dir) {
+			pipeline_list = &tp->pcm_info->capture_pipeline_list;
+			ret = tb_free_widgets_capture(tp, host, host);
+			if (ret < 0) {
+				fprintf(stderr, "failed to free widgets for capture PCM\n");
+				return ret;
+			}
+		} else {
+			pipeline_list = &tp->pcm_info->playback_pipeline_list;
+			ret = tb_free_widgets(tp, host, host);
+			if (ret < 0) {
+				fprintf(stderr, "failed to free widgets for playback PCM\n");
+				return ret;
+			}
+		}
+		for (i = 0; i < pipeline_list->count; i++) {
+			struct tplg_pipeline_info *pipe_info = pipeline_list->pipelines[i];
+
+			ret = tb_delete_pipeline(tp, pipe_info);
+			if (ret < 0)
+				return ret;
+		}
+	}
+
+	tp->instance_ids[SND_SOC_TPLG_DAPM_SCHEDULER] = 0;
+	return 0;
+}
+
+int tb_free_all_pipelines(struct testbench_prm *tp)
+{
+	tb_debug_print("freeing playback direction\n");
+	tb_free_pipelines(tp, SOF_IPC_STREAM_PLAYBACK);
+
+	tb_debug_print("freeing capture direction\n");
+	tb_free_pipelines(tp, SOF_IPC_STREAM_CAPTURE);
+	return 0;
+}
+
+void tb_free_topology(struct testbench_prm *tp)
+{
+	struct tplg_pcm_info *pcm_info;
+	struct tplg_comp_info *comp_info;
+	struct tplg_route_info *route_info;
+	struct tplg_pipeline_info *pipe_info;
+	struct tplg_context *ctx = &tp->tplg;
+	struct sof_ipc4_available_audio_format *available_fmts;
+	struct list_item *item, *_item;
+
+	list_for_item_safe(item, _item, &tp->pcm_list) {
+		pcm_info = container_of(item, struct tplg_pcm_info, item);
```

Comment:
rright, you have no concurrency, so this should be safe - you use the "safe" list iterator, so it's ok to free the current element, but you leave the freed elements on the list, so your list now contains invalid entries. You probable free the whole infrastructure after this or terminate the testbench completely, but just for correctness it would be better to delete each member before freeing it. Same in other "safe" loops below